### PR TITLE
Prioritize v2 option support

### PR DESCRIPTION
**Pull request by kunagpal**
Scope:
v1 normalization, v1 -> v2.x transformation

With this pull request:
Operations in the above defined scope will accept an option `prioritizeV2`, which will give preference to v2 style properties for the result of incoming v1 entities.

### CODE REVIEW

**Comment by kamalaknn** [comment]
@kunagpal we'd need the same `prioritizev2` option to be handled in `v1-v2-1` transformation as well.

File: test/unit/v1.0.0/converter-v1-to-v2.test.js [hunk 1]
```diff
@@ -493,14 +485,135 @@ describe('v1.0.0 to v2.0.0', function () {
...
-                            auth: null,
                             item: []
                         }]
                     });
                     done();
                 });
             });
         });
+
+        describe('prioritizeV2: true', function () {
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v2.test.js]: Can you also add tests for

1. missing legacy
2. missing non-legacy properties.
2. both legacy and non-legacy auth missing

tests for `prioritizeV2` for events and `preRequestScript` and `tests`

**Comment by kunagpal** [comment]
@kamalaknn That is already being done, as the v1 -> v2.1 inherits from the v1 -> v2. There are tests as well.

File: test/unit/v1.0.0/converter-v1-to-v2.test.js [hunk 2]
```diff
@@ -638,5 +751,115 @@ describe('v1.0.0 to v2.0.0', function () {
                 done();
             });
         });
+
+        describe('prioritizeV2: true', function () {
+            var options = {
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v2.test.js]: Could you add same tests mentioned above for events as well.

File: test/unit/v1.0.0/converter-v1-to-v21.test.js [hunk 3]
```diff
@@ -494,14 +486,140 @@ describe('v1.0.0 to v2.1.0', function () {
...
                             item: []
                         }]
                     });
                     done();
                 });
             });
         });
+
+        describe('prioritizeV2: true', function () {
+            var options = {
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v21.test.js]: Need all the same tests here, with and without `noDefaults`

File: test/unit/v1.0.0/converter-v1-to-v21.test.js [hunk 4]
```diff
@@ -494,14 +486,140 @@ describe('v1.0.0 to v2.1.0', function () {
...
+                inputVersion: '1.0.0',
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v21.test.js]: Need same additions as above.

File: test/unit/v1.0.0/normalize-v1.test.js [hunk 5]
```diff
@@ -605,6 +605,129 @@ describe('v1.0.0 normalization', function () {
                     });
                 });
             });
+
+            describe('prioritizeV2: true', function () {
+                var options = {
+                    normalizeVersion: '1.0.0',
```
Comment (kamalaknn) [inline test/unit/v1.0.0/normalize-v1.test.js]: Need the same tests for `noDefaults:true`

File: test/unit/v1.0.0/converter-v1-to-v2.test.js (hunk 1 above)
Comment (kamalaknn) [reply test/unit/v1.0.0/converter-v1-to-v2.test.js]: @kunagpal can we add the same tests for `prioritize:false`

**Comment by kunagpal** [comment]
@kamalaknn `prioritizeV2` is false by default, all other tests are effectively for that case.

File: lib/converters/v1.0.0/converter-v1-to-v2.js [hunk 6]
```diff
@@ -272,9 +272,12 @@ _.assign(Builders.prototype, {
      */
     auth: function (entityV1, options) {
         if (!entityV1) { return; }
-        if (util.notLegacy(entityV1, 'auth')) { return util.authArrayToMap(entityV1, options); }
-        if ((entityV1.currentHelper === null) || (entityV1.currentHelper === 'normal')) { return null; }
-        if (!entityV1.currentHelper) { return; }
+        if ((util.notLegacy(entityV1, 'auth') || this.options.prioritizeV2) && entityV1.auth) {
```
Comment (kamalaknn) [inline lib/converters/v1.0.0/converter-v1-to-v2.js]: @kunagpal could you flip this checks to 

```
if (prioritizeV2) {
  if (entityV1.auth) {}
  else if (entityV1.currentHelper) {}// or a more accurate legacy check
  else
}
else {
  if (entityV1.currentHelper) {}// or a more accurate legacy check
  else if (entityV1.auth) {}
  else
}
```

Comment (kunagpal) [reply lib/converters/v1.0.0/converter-v1-to-v2.js]: That makes the code a lot more verbose 😓

Comment (kamalaknn) [reply lib/converters/v1.0.0/converter-v1-to-v2.js]: Yeah but it makes it easy to reason about, and reduces possible bugs.

**Comment by kamalaknn** [comment]
@kunagpal I checked the tests for default options, `prioritizeV2:false` it does not have test for falsy and absent properties I guess.

File: lib/converters/v1.0.0/converter-v1-to-v2.js [hunk 7]
```diff
@@ -217,7 +217,7 @@ _.assign(Builders.prototype, {
         if (!entityV1) { return; }
 
         // events is treated as the source of truth in v1, so handle that first and bail out.
```
Comment (kamalaknn) [inline lib/converters/v1.0.0/converter-v1-to-v2.js]: This doc needs to be updated for `prioritizeV2`

File: lib/normalizers/v1.js [hunk 8]
```diff
@@ -55,17 +55,22 @@ _.assign(Builders.prototype, {
...
+        var auth,
+            params,
+            mapper,
+            currentHelper,
+            helperAttributes,
+            prioritizeV2 = this.options.prioritizeV2;
+
         // Handle the noauth/normal case for new and legacy attributes first, in that order
-        if (util.notLegacy(entityV1, 'auth')) {
+        if ((util.notLegacy(entityV1, 'auth') || prioritizeV2) && (entityV1.auth || !prioritizeV2)) {
```
Comment (kamalaknn) [inline lib/normalizers/v1.js]: This particular condition is completely unreadable. This could have been,

`if ((entityV1.auth && prioritizeV2) || util.notLegacy(entityV1, 'auth'))`

File: lib/normalizers/v1.js [hunk 9]
```diff
@@ -55,17 +55,22 @@ _.assign(Builders.prototype, {
...
         if (!entityV1) { return; }
+
+        var auth,
+            params,
+            mapper,
+            currentHelper,
+            helperAttributes,
+            prioritizeV2 = this.options.prioritizeV2;
+
         // Handle the noauth/normal case for new and legacy attributes first, in that order
```
Comment (kamalaknn) [inline lib/normalizers/v1.js]: Doc needs to be updated

File: test/unit/v1.0.0/converter-v1-to-v2.test.js [hunk 10]
```diff
@@ -493,14 +485,309 @@ describe('v1.0.0 to v2.0.0', function () {
...
+                                basic: { username: 'postman', password: 'secret', showPassword: false }
+                            }
+                        },
+                        response: []
+                    });
+                    done();
+                });
+            });
+
+            it('should discard auth creation if both: legacy and new attributes are missing', function (done) {
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v2.test.js]: Need to add a test case for `prioritizeV2:false` and `auth` present and `legacy` properties missing

File: test/unit/v1.0.0/converter-v1-to-v2.test.js [hunk 11]
```diff
@@ -493,14 +485,309 @@ describe('v1.0.0 to v2.0.0', function () {
...
+                        response: []
+                    });
+                    done();
+                });
+            });
+
+            it('should discard auth if both: legacy is normal and new attributes are missing', function (done) {
+                var source = {
+                    id: '27ad5d23-f158-41e2-900d-4f81e62c0a1c',
+                    currentHelper: 'normal'
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v2.test.js]: Need test for `currentHelper:null` as well.

File: test/unit/v1.0.0/converter-v1-to-v2.test.js [hunk 12]
```diff
@@ -493,14 +485,309 @@ describe('v1.0.0 to v2.0.0', function () {
...
+                        response: []
+                    });
+                    done();
+                });
+            });
+
+            it('should discard auth creation if both: legacy is normal and new attributes is falsy', function (done) {
+                var source = {
+                    id: '27ad5d23-f158-41e2-900d-4f81e62c0a1c',
+                    currentHelper: 'normal',
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v2.test.js]: Need test for `currentHelper:null` as well.

File: test/unit/v1.0.0/converter-v1-to-v2.test.js [hunk 13]
```diff
@@ -493,14 +485,309 @@ describe('v1.0.0 to v2.0.0', function () {
...
+                            response: []
+                        });
+                        done();
+                    });
+                });
+
+                it('should discard auth if both: legacy is normal and new attributes are missing', function (done) {
+                    var source = {
+                        id: '27ad5d23-f158-41e2-900d-4f81e62c0a1c',
+                        currentHelper: 'normal'
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v2.test.js]: Need test for `currentHelper:null`

File: test/unit/v1.0.0/converter-v1-to-v2.test.js [hunk 14]
```diff
@@ -638,5 +925,327 @@ describe('v1.0.0 to v2.0.0', function () {
...
         });
+
+        describe('with missing properties', function () {
+            var options = {
+                inputVersion: '1.0.0',
+                outputVersion: '2.0.0',
+                retainIds: true
+            };
+
+            it('should handle missing preRequestScript and tests correctly', function (done) {
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v2.test.js]: This is the same as the test above.

File: test/unit/v1.0.0/converter-v1-to-v2.test.js [hunk 15]
```diff
@@ -638,5 +925,327 @@ describe('v1.0.0 to v2.0.0', function () {
...
+                        }]
+                    });
+                    done();
+                });
+            });
+
+            it('should discard event from the result if both legacy and current attributes are empty', function (done) {
+                var source = {
+                    id: '27ad5d23-f158-41e2-900d-4f81e62c0a1c',
+                    preRequestScript: null,
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v2.test.js]: Need tests for `preRequestScript:''`

File: test/unit/v1.0.0/converter-v1-to-v2.test.js (hunk 12 above)
Comment (kunagpal) [reply test/unit/v1.0.0/converter-v1-to-v2.test.js]: This one already exists.

File: test/unit/v1.0.0/converter-v1-to-v2.test.js (hunk 13 above)
Comment (kunagpal) [reply test/unit/v1.0.0/converter-v1-to-v2.test.js]: Already exists.

File: test/unit/v1.0.0/converter-v1-to-v2.test.js (hunk 14 above)
Comment (kunagpal) [reply test/unit/v1.0.0/converter-v1-to-v2.test.js]: Fixed.

File: test/unit/v1.0.0/converter-v1-to-v21.test.js [hunk 16]
```diff
@@ -494,14 +486,324 @@ describe('v1.0.0 to v2.1.0', function () {
...
+                        response: []
+                    });
+                    done();
+                });
+            });
+
+            it('should discard auth if both: legacy is normal and new attributes are missing', function (done) {
+                var source = {
+                    id: '27ad5d23-f158-41e2-900d-4f81e62c0a1c',
+                    currentHelper: 'normal'
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v21.test.js]: Need test for `currentHelper: null`

Comment (kunagpal) [reply test/unit/v1.0.0/converter-v1-to-v21.test.js]: Added

File: test/unit/v1.0.0/converter-v1-to-v21.test.js [hunk 17]
```diff
@@ -494,14 +486,324 @@ describe('v1.0.0 to v2.1.0', function () {
...
-                            auth: null,
                             item: []
                         }]
                     });
                     done();
                 });
             });
         });
+
+        describe('with missing properties', function () {
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v21.test.js]: Need test for missing `legacy` properties with `.auth`

File: test/unit/v1.0.0/normalize-v1.test.js [hunk 18]
```diff
@@ -605,6 +605,194 @@ describe('v1.0.0 normalization', function () {
...
                 });
             });
+
+            describe('with missing properties', function () {
+                var options = {
+                    noDefaults: true,
+                    normalizeVersion: '1.0.0'
+                };
+
+                it('should fall back to legacy properties if auth is missing', function (done) {
```
Comment (kamalaknn) [inline test/unit/v1.0.0/normalize-v1.test.js]: Need an inverse missing test.

Comment (kunagpal) [reply test/unit/v1.0.0/normalize-v1.test.js]: Exists

File: test/unit/v1.0.0/converter-v1-to-v21.test.js (hunk 17 above)
Comment (kunagpal) [reply test/unit/v1.0.0/converter-v1-to-v21.test.js]: Exists
